### PR TITLE
Several cppcheck related fixes

### PR DESCRIPTION
**Pull request by rouault**
* if (this != &other) in copy constructors
* missing initializations of member variables
* missing copy of member variables in copy constructor
* a couple other issues

### CODE REVIEW

File: src/analysis/interpolation/DualEdgeTriangulation.cpp [hunk 1]
```diff
@@ -2915,25 +2915,20 @@ bool DualEdgeTriangulation::swapEdge( double x, double y )
 QList<int> *DualEdgeTriangulation::getPointsAroundEdge( double x, double y )
 {
   QgsPoint p( x, y, 0 );
-  int p1, p2, p3, p4;
-  int edge1 = baseEdgeOfTriangle( p );
+  const int int edge1 = baseEdgeOfTriangle( p );
```
Comment (nyalldawson) [inline src/analysis/interpolation/DualEdgeTriangulation.cpp:2918]:
```suggestion
  const int edge1 = baseEdgeOfTriangle( p );
```

File: src/core/geometry/qgsgeometrycollection.cpp [hunk 2]
```diff
@@ -36,7 +36,8 @@ QgsGeometryCollection::QgsGeometryCollection()
   mWkbType = QgsWkbTypes::GeometryCollection;
 }
 
-QgsGeometryCollection::QgsGeometryCollection( const QgsGeometryCollection &c ): QgsAbstractGeometry( c )
+QgsGeometryCollection::QgsGeometryCollection( const QgsGeometryCollection &c ):
+  QgsAbstractGeometry( c ), mHasCachedValidity( false ), mValidityFailureReason()
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometrycollection.cpp:40]: we could copy the cached validity flag/reason here too

File: src/analysis/interpolation/DualEdgeTriangulation.cpp (hunk 1 above)
Comment (rouault) [reply src/analysis/interpolation/DualEdgeTriangulation.cpp]: fixed

File: src/core/geometry/qgsgeometrycollection.cpp (hunk 2 above)
Comment (rouault) [reply src/core/geometry/qgsgeometrycollection.cpp]: makes sense. Done

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**